networking_aci/extensions/acioperations.py

```diff
@@ -0,0 +1,219 @@
+
+        return hostgroup
+
+
+class HostgroupModeController(wsgi.Controller):
+    def __init__(self, plugin, db, rpc_notifier):
+        super(HostgroupModeController, self).__init__()
+        self.plugin = plugin
+        self.db = db
+        self.rpc_notifier = rpc_notifier
+
+    def index(self, request, **kwargs):
+        check_cloud_admin(request)
+        ctx = context.get_admin_context()
+        return dict(hostgroup_modes=self.db.get_hostgroup_modes(ctx))
+
+    def show(self, request, **kwargs):
+        check_cloud_admin(request)
+        hostgroup_name = kwargs.pop('id')
+        ctx = context.get_admin_context()
+        return dict(hostgroup_name=hostgroup_name, hostgroup_mode=self._get_mode(ctx, hostgroup_name))
+
+    def update(self, request, **kwargs):
+        check_cloud_admin(request)
+        hostgroup_name = kwargs.pop('id')
+        new_mode = kwargs.get("body", {}).get("mode")
+        if new_mode not in (aci_const.MODE_INFRA, aci_const.MODE_BAREMETAL):
+            raise web_exc.HTTPBadRequest("Specify a mode that is either {} or {} (given mode was '{}')"
+                                         .format(aci_const.MODE_INFRA, aci_const.MODE_BAREMETAL, new_mode))
+
+        ctx = context.get_admin_context()
+        curr_mode = self._get_mode(ctx, hostgroup_name)
+        if curr_mode == new_mode:
+            raise web_exc.HTTPBadRequest("{} is already in {} mode".format(hostgroup_name, new_mode))
+
+        # check if there are still  on this hostgroup
+        hg_config = ACI_CONFIG.get_hostgroup(hostgroup_name)
+        if not hg_config:
+            raise web_exc.HTTPBadRequest("Hostgroup {} has no config associated with it".format(hostgroup_name))
+
+        for segment_id in self.db.get_segment_ids_by_physnet(ctx, hg_config['physical_network']):
+            hosts = self.db.get_hosts_on_segment(ctx, segment_id)
+            for host in hg_config['hosts']:
+                if host in hosts:
+                    LOG.error("Denied API request to switch hostgroup %s to %s mode, host %s still bound in segment %s",
+                              hostgroup_name, new_mode, host, segment_id)
+                    raise web_exc.HTTPConflict("Could not switch, there is still a portbinding for host {} "
+                                               "present in segment {}"
+                                               .format(host, segment_id))
+
+        # cleanup old baremetal objects when switching away from baremetal mode
+        baremetal_bindings_cleaned = False
+        if hg_config['hostgroup_mode'] == aci_const.MODE_BAREMETAL:
+            LOG.info("Cleaning baremetal objects for %s from device", hostgroup_name)
+            try:
+                self.rpc_notifier.clean_baremetal_objects(hg_config)
+                baremetal_bindings_cleaned = True
+            except Exception:
+                LOG.exception("Could not clean baremetal bindings from ACI, cleaning will have to be done manually")
```

Comment:
The client will not see that something went wrong. Can we provide the info somehow in an additional field? Might help in automating this via ansible later on.
Or does this need domain-knowledge in the automation, that we need to check `baremetal_bindings_cleaned` when setting away from `MODE_BAREMETAL`? Fine, too.

Reply:
Jup, this is hard domain knowledge. The flag `baremetal_bindings_cleaned` is only interesting in the switchover case from baremetal to infrastructure mode (else there are no bindings to clean). If the API returns false then there was a cleanup error.